HIP-3.md

```diff
@@ -0,0 +1,189 @@
+     *   and required to be at least the msg.value. Any overpayment is sent back
+     *   to the `_refundAddress`.
+     *   Emits the `GasPayment` event with _handleGas gas and IsmType.Default.
+     * @param _messageId The message ID.
+     * @param _destinationDomain The destination domain of the message.
+     * @param _handleGas The amount of gas required by the recipient's `handle` function.
+     * @param _refundAddress The address to refund any msg.value overpayment to.
+     */
+    function payGasForDefaultIsm(
+        bytes32 _messageId,
+        uint32 _destinationDomain,
+        uint256 _handleGas,
+        address _refundAddress
+    ) external payable;
+
+    /**
+     * @notice Accepts payment for a message that uses a non-default ISM on the
+     *   destination. The `_ismAndHandleGas` should cover ISM metadata calldata gas,
+     *   ISM.verify gas, and gas required by the recipient's `handle` function.
+     *   The gas required for other operations involving the processing of the
+     *   message is read from storage and is added to `_handleGas` to calculate
+     *   the total estimated amount of destination gas that's required.
+     *   The stored value includes intrinsic gas, approximated message calldata
+     *   gas, and any overhead gas incurred by the Mailbox contract.
+     *   Using the total expected destination gas usage, the destination gas price
+     *   (from an oracle), and the destination chain native token quoted in the origin
+     *   chain native token (from an oracle), the required fee payment is calculated
+     *   and required to be at least the msg.value. Any overpayment is sent back
+     *   to the `_refundAddress`.
+     *   Emits the `GasPayment` event with _ismAndHandleGas gas and IsmType.NonDefault.
+     * @param _messageId The message ID.
+     * @param _destinationDomain The destination domain of the message.
+     * @param _ismAndHandleGas The amount of gas to cover ISM metadata calldata,
+     *   ISM.verify gas, and gas required by the recipient's `handle` function.
+     * @param _refundAddress The address to refund any msg.value overpayment to.
+     */
+    function payGasForNonDefaultIsm(
+        bytes32 _messageId,
+        uint32 _destinationDomain,
+        uint256 _ismAndHandleGas,
+        address _refundAddress
+    ) external payable;
+```
+
+The off-chain relayer will index `GasPayment` events and relate them to dispatched messages based off their message ID.
+
+For the default ISM case, the relayer will create the process transaction with a gas limit that results in at least `_handleGas` gas able to be used by the recipient’s `handle` function. The total gas limit of the transaction can be found by the relayer estimating all other gas costs (i.e. intrinsic gas, calldata gas including the ISM metadata and message body, ISM.verify gas, and Mailbox overhead gas) and adding this to the `_handleGas`.
```

Comment:
Doesn't the IGP will need to calculate all of this also?

Reply:
Yeah so the idea is to have that be read from storage

The motivation behind separating default ISM from not is that the default ISM can and will change frequently and user's shouldn't need to take action based off that